### PR TITLE
Add type hints to private properties

### PR DESCRIPTION
**Pull request by fabpot**

### CODE REVIEW

File: src/Extension/StagingExtension.php [hunk 1]
```diff
@@ -26,11 +26,26 @@
...
+    /**
+     * @var array<string, TwigFunction>
+     */
+    private array $functions = [];
+    /**
+     * @var array<string, TwigFilter>
+     */
+    private array $filters = [];
+    /**
+     * @var array<int, NodeVisitorInterface>
```
Comment (GromNaN) [inline src/Extension/StagingExtension.php]:
```suggestion
     * @var list<NodeVisitorInterface>
```

File: src/ExtensionSet.php [hunk 2]
```diff
@@ -27,28 +27,37 @@
...
     private $staging;
-    private $parsers;
-    private $visitors;
+    /**
+     * @var TokenParserInterface[]
+     */
+    private array $parsers;
+    /**
+     * @var NodeVisitorInterface[]
+     */
```
Comment (GromNaN) [inline src/ExtensionSet.php:43]: The other docblocks of this file are on a single line.
```suggestion
     /** @var list<NodeVisitorInterface> */
```

File: src/ExtensionSet.php [hunk 3]
```diff
@@ -27,28 +27,37 @@
...
+    private array $extensions;
+    private bool $initialized = false;
+    private bool $runtimeInitialized = false;
     private $staging;
-    private $parsers;
-    private $visitors;
+    /**
+     * @var TokenParserInterface[]
+     */
```
Comment (GromNaN) [inline src/ExtensionSet.php:39]:
```suggestion
    /** @var array<string, TokenParserInterface> */
```

File: src/ExtensionSet.php [hunk 4]
```diff
@@ -27,28 +27,37 @@
  */
 final class ExtensionSet
 {
-    private $extensions;
-    private $initialized = false;
-    private $runtimeInitialized = false;
+    /**
+     * @var ExtensionInterface[]
+     */
```
Comment (GromNaN) [inline src/ExtensionSet.php:32]:
```suggestion
    /** @var array<class-string<ExtensionInterface>, ExtensionInterface> */
```

File: src/Lexer.php [hunk 5]
```diff
@@ -19,23 +19,38 @@
...
-    private $options;
-    private $regexes;
-    private $position;
-    private $positions;
-    private $currentVarBlockLine;
+    private bool $isInitialized = false;
+
+    /**
+     * @var Token[]
+     */
```
Comment (GromNaN) [inline src/Lexer.php]:
```suggestion
    /** @var list<Token> */
```

**Review by GromNaN — COMMENTED** [review]
You can use `list<Type>` when items are only appended with `$var[]`.

File: composer.json [hunk 6]
```diff
@@ -32,7 +32,8 @@
     },
     "require-dev": {
         "symfony/phpunit-bridge": "^5.4.9|^6.3|^7.0",
-        "psr/container": "^1.0|^2.0"
+        "psr/container": "^1.0|^2.0",
+        "phpstan/phpstan": "1.11.x-dev"
```
Comment (stof) [inline composer.json]: why requiring a dev version instead of the released version ?

File: src/Compiler.php [hunk 7]
```diff
@@ -19,14 +19,14 @@
...
 {
-    private $lastLine;
-    private $source;
-    private $indentation;
-    private $env;
-    private $debugInfo = [];
-    private $sourceOffset;
-    private $sourceLine;
-    private $varNameSalt = 0;
+    private ?int $lastLine;
```
Comment (stof) [inline src/Compiler.php:22]: shouldn't this use ` = null` to initialize it ?

File: src/Environment.php [hunk 8]
```diff
@@ -429,7 +441,7 @@ public function isTemplateFresh(string $name, int $time): bool
      * @throws LoaderError When none of the templates can be found
      * @throws SyntaxError When an error occurred during compilation
      */
-    public function resolveTemplate($names): TemplateWrapper
+    public function resolveTemplate($names): TemplateWrapper|Template
```
Comment (stof) [inline src/Environment.php]: should we actually support passing a Template in the array of names ? And if yes, shouldn't we wrap it when returning it ?

File: src/Extension/EscaperExtension.php [hunk 9]
```diff
@@ -110,7 +113,7 @@ public function setEscaper($strategy, callable $callable)
     /**
      * Gets all defined escapers.
      *
-     * @return callable[] An array of escapers
+     * @return array<string, callable> An array of escapers
```
Comment (stof) [inline src/Extension/EscaperExtension.php:116]: I suggest using `callable(Environment, string, string): string` to describe the signature of the callable.

File: src/RuntimeLoader/FactoryRuntimeLoader.php [hunk 10]
```diff
@@ -18,7 +18,10 @@
  */
 class FactoryRuntimeLoader implements RuntimeLoaderInterface
 {
-    private $map;
+    /**
+     * @var callable[]
```
Comment (stof) [inline src/RuntimeLoader/FactoryRuntimeLoader.php]: `array<string, callable(): object>` (and same should be added on the constructor)

File: composer.json (hunk 6 above)
Comment (stof) [reply composer.json]: also, a config file should probably be added for it if you start using it.

Comment (fabpot) [reply composer.json]: My bad, removed.

File: src/Compiler.php (hunk 7 above)
Comment (fabpot) [reply src/Compiler.php:22]: I've removed all default values here and called `reset()` in the constructor instead.

File: src/Extension/EscaperExtension.php (hunk 9 above)
Comment (fabpot) [reply src/Extension/EscaperExtension.php:116]: Done

File: src/RuntimeLoader/FactoryRuntimeLoader.php (hunk 10 above)
Comment (fabpot) [reply src/RuntimeLoader/FactoryRuntimeLoader.php]: Done

File: src/Extension/StagingExtension.php (hunk 1 above)
Comment (fabpot) [reply src/Extension/StagingExtension.php]: Done

File: src/ExtensionSet.php (hunk 2 above)
Comment (fabpot) [reply src/ExtensionSet.php:43]: I've moved the other ones to multiplines to be consistent with all the other files.

File: src/Lexer.php (hunk 5 above)
Comment (fabpot) [reply src/Lexer.php]: Done

File: src/Environment.php (hunk 8 above)
Comment (fabpot) [reply src/Environment.php]: Removed this and wrapped Template instances instead.

**Pull request merged**